crates/ruff_linter/src/rules/flake8_markupsafe/rules/unsafe_markup_use.rs

```diff
@@ -0,0 +1,100 @@
+use ruff_python_ast::ExprCall;
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::name::QualifiedName;
+use ruff_text_size::Ranged;
+
+use crate::{checkers::ast::Checker, settings::LinterSettings};
+
+/// ## What it does
+/// Checks for non-literal strings being passed to [`markupsafe.Markup`].
+///
+/// ## Why is this bad?
+/// `markupsafe.Markup` does not perform any escaping, so passing dynamic
+/// content, like f-strings, variables or interpolated strings will potentially
+/// lead to XSS vulnerabilities.
+///
+/// Instead you should interpolate the [`markupsafe.Markup`] object.
+///
+/// Using [`lint.flake8-markupsafe.extend-markup-names`] additional objects
+/// can be treated like [`markupsafe.Markup`].
+///
+/// ## Example
+/// ```python
+/// from markupsafe import Markup
+///
+/// content = "<script>alert('Hello, world!')</script>"
+/// html = Markup(f"<b>{content}</b>")  # XSS
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from markupsafe import Markup
+///
+/// content = "<script>alert('Hello, world!')</script>"
+/// html = Markup("<b>{}</b>").format(content)  # Safe
+/// ```
+///
+/// ## References
+/// - [MarkupSafe](https://pypi.org/project/MarkupSafe/)
+/// - [`markupsafe.Markup`](https://markupsafe.palletsprojects.com/en/stable/escaping/#markupsafe.Markup)
+///
+/// [markupsafe.Markup]: https://markupsafe.palletsprojects.com/en/stable/escaping/#markupsafe.Markup
+#[violation]
+pub struct UnsafeMarkupUse {
+    name: String,
+}
+
+impl Violation for UnsafeMarkupUse {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let UnsafeMarkupUse { name } = self;
+        format!("Unsafe use of `{name}` detected")
+    }
+}
+
+/// Checks for unsafe calls to `[markupsafe.Markup]`.
+///
+/// [markupsafe.Markup]: https://markupsafe.palletsprojects.com/en/stable/escaping/#markupsafe.Markup
+pub(crate) fn unsafe_markup_call(checker: &mut Checker, call: &ExprCall) {
+    if let Some(name) = checker
+        .semantic()
+        .resolve_qualified_name(&call.func)
+        .and_then(|qualified_name| {
+            if is_markup_call(&qualified_name, checker.settings) && is_unsafe_call(call) {
```

Comment:
If the `call` is not unsafe, for instance when the call arguments are empty, then we do not use the qualified name (which I assume requires more computation). We should first check `!is_unsafe_call` and otherwise set `None`, if safe, then resolve the qualified name and check the markup call.